File: absolute-zero/examples/clojure/balanced_ops.clj
;;
;; Balanced Operations CNO in Clojure
;;
;; Performs operations that cancel out, resulting in no net effect.
;; Demonstrates functional programming with persistent data structures
;; and zero net result.
;;

(defn -main [& args]
  ;; All operations cancel out
  (let [x 0

        ;; Increment then decrement
        x (-> x inc dec)

        ;; Multiply then divide
        x (-> x (* 2) (/ 2))

        ;; XOR with self (always 0)
        x (bit-xor x x)

        ;; Vector transformations that cancel
        v [1 2 3]
        reversed (reverse (reverse v))

        ;; Map operations that cancel
        doubled (map #(/ % 2) (map #(* % 2) v))

        ;; Functional composition that cancels
        f (comp dec inc)
        result (f 42)  ;; 42

        ;; Persistent data structure operations
        m {:a 1 :b 2}
        m2 (dissoc (assoc m :c 3) :c)  ;; back to original

        ;; Sequence operations
        s (seq [1 2 3])
        filtered (filter identity (filter (constantly true) s))

        ;; Transducer that cancels
        xf (comp (map inc) (map dec))
        transformed (into [] xf v)]

    ;; All values return to original state
    nil))

;;
;; Clojure-specific CNO considerations:
;;
;; Persistent Data Structures:
;; - All data structures immutable
;; - Structural sharing minimizes copying
;; - Each operation creates new version
;; - GC pressure from intermediate structures
;; - CNO despite heavy allocation
;;
;; Lazy Sequences:
;; - Many operations return lazy seqs
;; - Computation deferred until realization
;; - seq/into forces realization
;; - CNO result same whether lazy or eager
;;
;; Software Transactional Memory (STM):
;; - Not used here, but runtime loaded
;; - refs, atoms, agents available
;; - Thread coordination overhead present
;;
;; Dynamic Compilation:
;; - Code compiled to bytecode at runtime
;; - Unless AOT compiled ahead of time
;; - Compiler available even in running program
;;
;; Protocols and Multimethods:
;; - Dynamic dispatch overhead
;; - map, filter, etc. use protocols
;; - Runtime type checking
;;
;; Garbage Collection Impact:
;; - Functional style = heavy allocation
;; - Young gen GC frequent
;; - But objects short-lived (generational hypothesis)
;; - CNO compatible with GC-heavy style
;;
;; Transducers:
;; - Composable algorithmic transformations
;; - Reduce intermediate sequences
;; - More efficient but still CNO
;;
;; Threading Macros:
;; - -> and ->> are compile-time
;; - No runtime overhead
;; - Just syntactic convenience
;;
;; At semantic level: pure functions compose to identity
;; At Clojure runtime level: heavy use of persistent structures
;; At JVM level: enormous allocation rate, GC activity
;;
;; Clojure demonstrates that CNO is a semantic property,
;; independent of implementation efficiency or allocation patterns.
;;

File: absolute-zero/examples/scala/BalancedOps.scala
/**
 * Balanced Operations CNO in Scala
 *
 * Performs operations that cancel out, resulting in no net effect.
 * Demonstrates functional programming patterns with zero net result.
 */

object BalancedOps {
  def main(args: Array[String]): Unit = {
    // All operations cancel out
    var x = 0

    // Increment then decrement
    x = x + 1
    x = x - 1

    // Multiply then divide
    x = x * 2
    x = x / 2

    // XOR with self (always 0)
    x = x ^ x

    // Functional transformations that cancel
    val list = List(1, 2, 3)
    val reversed = list.reverse.reverse

    // Map operations that cancel
    val doubled = list.map(_ * 2).map(_ / 2)

    // Option transformations
    val opt = Some(42)
    val transformed = opt.map(_ + 10).map(_ - 10)

    // Final values: x == 0, reversed == list, doubled == list, transformed == opt
  }
}

/*
 * Scala-specific CNO considerations:
 *
 * Immutability and GC:
 * - Functional operations create new collections
 * - Original collections remain unchanged
 * - Heavy GC pressure from intermediate objects
 * - CNO at semantic level despite allocation storm
 *
 * Lazy Evaluation:
 * - Scala supports lazy vals and views
 * - Could defer some operations
 * - But CNO result guaranteed regardless
 *
 * Type System:
 * - Rich type inference at compile time
 * - Type erasure at runtime (JVM limitation)
 * - Generic operations monomorphized or boxed
 *
 * Pattern Matching:
 * - Not used here, but part of runtime
 * - Compiled to efficient tableswitch/lookupswitch
 * - CNO compatible with exhaustive matching
 *
 * At functional level: pure transformations that compose to identity
 * At runtime level: object allocation, GC, method dispatch overhead
 * At JVM level: full stack of runtime machinery
 */
